homeassistant/components/mqtt/discovery.py

```diff
@@ -275,32 +275,34 @@ def async_process_discovery_payload(
 
         _LOGGER.debug("Process discovery payload %s", payload)
         discovery_hash = (component, discovery_id)
-        if discovery_hash in mqtt_data.discovery_already_discovered or payload:
+
+        already_discovered = discovery_hash in mqtt_data.discovery_already_discovered
+        if (
+            already_discovered or payload
+        ) and discovery_hash not in mqtt_data.discovery_pending_discovered:
+            discovery_pending_discovered = mqtt_data.discovery_pending_discovered
 
             @callback
             def discovery_done(_: Any) -> None:
-                pending = mqtt_data.discovery_pending_discovered[discovery_hash][
-                    "pending"
-                ]
+                pending = discovery_pending_discovered[discovery_hash]["pending"]
                 _LOGGER.debug("Pending discovery for %s: %s", discovery_hash, pending)
                 if not pending:
-                    mqtt_data.discovery_pending_discovered[discovery_hash]["unsub"]()
-                    mqtt_data.discovery_pending_discovered.pop(discovery_hash)
+                    discovery_pending_discovered[discovery_hash]["unsub"]()
+                    discovery_pending_discovered.pop(discovery_hash)
                 else:
                     payload = pending.pop()
                     async_process_discovery_payload(component, discovery_id, payload)
 
-            if discovery_hash not in mqtt_data.discovery_pending_discovered:
```

Comment:
The functional change here is to move this if into the if above so we don't enter this block and create the closure function above if we are just going to throw it away